File: 102. Binary Tree Level Order Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int> > res;
        if (!root) return res;
        res.push_back(vector<int>{});
        queue<TreeNode*> q;
        int cur = 1, nxt = 0;
        q.push(root);
        while (!q.empty()) {
            TreeNode* tp = q.front();
            q.pop();
            cur--;
            res.back().push_back(tp->val);
            if (tp->left) {
                q.push(tp->left);
                nxt++;
            }
            if (tp->right) {
                q.push(tp->right);
                nxt++;
            }
            if (cur == 0) {
                if (nxt == 0) break;
                cur = nxt;
                nxt = 0;
                res.push_back(vector<int>{});
            }
        }
        return res;
    }
};

File: java/102. Binary Tree Level Order Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
// O(n) time, O(n) space
class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> res = new ArrayList<>();
        if (root == null) {
            return res;
        }
        int count = 1, nextCount = 0;
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        res.add(new ArrayList<>());
        while (!queue.isEmpty()) {
            TreeNode cur = queue.poll();
            --count;
            res.get(res.size() - 1).add(cur.val);
            if (cur.left != null) {
                queue.offer(cur.left);
                nextCount++;
            }
            if (cur.right != null) {
                queue.offer(cur.right);
                nextCount++;
            }
            if (count == 0 && nextCount != 0) {
                count = nextCount;
                nextCount = 0;
                res.add(new ArrayList<>());
            }
        }
        return res;
    }
}
